### DISCUSSION TITLE
How to use SelectedValueBinding on combo box with Compiled Binding

### DISCUSSION THREAD

**Discussion by nixtar** [AvaloniaUI/Avalonia#14760] [Q&A] [answered]
Hello,

I'm trying to implement compiled bindings on an app I'm working on so I can enable full trimming.

I have a combo box with its ItemsSource bound to an array of objects in the ViewModel, SelectedValue is bound to a string in the ViewModel and SelectedValueBinding is set to a string property on the object.

EG:
```c#
public class User
{
    public string FirstName { get; set; }
    
    public string FullName { get; set; }
}

public class ViewModel
{
    public User[] Users { get; set; }
    
    public string SelectedUser { get; set; }
}
```

```xaml
<ComboBox ItemsSource="{Binding Users}"
          SelectedValue="{Binding Path=SelectedUser}"
          SelectedValueBinding="{Binding FullName}">
  <ComboBox.ItemTemplate>
    <DataTemplate>
      <TextBlock Text="{Binding FirstName}"
          x:DataType="business:User"/>
    </DataTemplate>
  </ComboBox.ItemTemplate>
</ComboBox>
```

This currently results in an error at compile time `Error AVLN:0004 Avalonia: Unable to resolve property or method of name 'FullName' on type 'ViewModels.MainViewModel'. Line 49, position 21.`

I've tried setting the DataType eg: `SelectedValueBinding="{Binding FullName, DataType={x:Type business:User}}"` however this fails with error: `Unable to resolve field or property 'FullName' in data context of type 'Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindingExtension'`

Is this a supported configuration and if so, how can I achieve this?

**Comment by maxkatz6** [comment]
SelectedValueBinding binding has `[InheritDataTypeFromItems(nameof(ItemsSource))]` attribute on it, so compiled should be able to find DataType from Users collection. In your code snipped you don't need any of these DataType definitions, as everything should be inherited from the ItemsSource, including Itemtemplate. 

`{Binding FullName, DataType=business:User}` should also work. I have never seen `Unable to resolve field or property 'FullName' in data context of type 'Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindingExtension'` error though.

**Reply by maxkatz6** [reply, marked as answer]
By chance, can you try to change Users type to IList or ObservableCollection, instead of Array?

**Reply by nixtar** [reply]
Ahh perfect! IList fixed it! ❤️

**Reply by maxkatz6** [reply]
It still seems to be a bug, that can be fixed. Arrays should also work.

**Reply by nixtar** [reply]
Happy to log this as an issue if you want.

**Comment by vpenades** [comment]
I have a somewhat similar question, which matches exactly the title of this thread, but it may be different under the hood (so sorry for stealing the discussion)

I the model view I have a plain string property:

```c:
public string SelectedValue { get; set; }
```

and I want to create a combobox with a twist; for every known value, I want to set a different _display value_ , the reason is because the display values need to be localized.

The xaml may look like this:

```xml

<ComboBox SelectedValue="{Binding SelectedValue}"
          SelectedValueBinding="{Binding DataContext}">

  <ComboBoxItem DataContext="LOW" Content="low value" />
  <ComboBoxItem DataContext="MEDIUM" Content="medium value" />
  <ComboBoxItem DataContext="HIGH" Content="high value" />
</ComboBox>
```

In WPF I was able to do something that worked, using SelectedValuePath, now SelectedValueBinding.

Essentially I want to use ComboBoxItem as a Key-Value dictionary, where the Key (DataContext) is the actual value used for binding, and the Value(Content) is used for display, which is a critical feature if you want the displayed text to be localizable in xaml.

I think that this works when using reflection binding (because as I understand it, SelectedValueBinding binds to the item of the collection).

But when using CompiledBindings it is required to set the x:DataType ... and here's the catch:

if the x:DataType is set to "ComboBoxItem" to make SelectedValueBinding work, SelectedValue stops working

if the x:DataType is set to "string" to match the Model's SelectedValue,  then SelectedValueBinding stops working because it's unable to find DataContext in "string".

I think this is not an uncommon case, because otherwise, using fixed collections of ComboBoxItems becomes quite useless when using display text.

Even [Avalonia's own documentation](https://docs.avaloniaui.net/docs/reference/controls/combobox) sneakily avoid the issue of SelectedValue + SelectedValueBinding by using the SelectedIndex instead.

I don't know which would be the right solution, having a second x:DataType for SelectedValueBinding woud be wild ... so maybe a special use case for ComboBoxItem in these scenarios?
